File: LeetCode Problems/1 to 100/14. Longest Common Prefix/14_Longest_Common_Prefix.java
public class Solution {
    static class Node {
        char info;
        int cnt;
        Node[] list;

        public Node(char info) {
            this.info = info;
            this.cnt = 0;
            this.list = new Node[26];
        }
    }

    public Node newNode(char info) {
        return new Node(info);
    }

    public void insert(Node head, String s) {
        Node temp = head;
        for (int i = 0; i < s.length(); i++) {
            int index = s.charAt(i) - 'a';
            if (temp.list[index] == null) {
                temp.list[index] = newNode(s.charAt(i));
            }
            temp = temp.list[index];
        }
        temp.cnt = 1;
    }

    public void display(Node root) {
        if (root == null) {
            return;
        }
        Node temp = root;
        System.out.println(temp.info + " " + temp.cnt);
        for (int i = 0; i < 26; i++) {
            display(temp.list[i]);
        }
    }

    public int countChild(Node temp, int[] index) {
        int cnt = 0;
        for (int i = 0; i < 26; i++) {
            if (temp.list[i] != null) {
                cnt++;
                index[0] = i;
            }
        }
        return cnt;
    }

    public String longestCommonPrefix(String[] strs) {
        Node head = newNode('\0');
        for (String str : strs) {
            insert(head, str);
        }
        Node curr = head;
        StringBuilder lcp = new StringBuilder();
        int[] index = new int[1];
        while (curr.cnt == 0 && countChild(curr, index) == 1) {
            curr = curr.list[index[0]];
            lcp.append((char) ('a' + index[0]));
        }
        return lcp.toString();
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        String[] strs = {"flower", "flow", "flight"};
        String lcp = solution.longestCommonPrefix(strs);
        System.out.println("Longest Common Prefix: " + lcp);
    }
}

File: LeetCode Problems/1 to 100/14. Longest Common Prefix/14. Longest Common Prefix.cpp
class Solution {
public:
    struct node {
        char info;
        int cnt;
        node *list[26];
    };
    node *newNode(char info) {
        node *temp = new node();
        temp->cnt = 0;
        temp->info = info;
        return temp;
    }
    void insert(node *&head, string s) {
        node *temp = head;
        for (int i = 0; i < s.size(); i++) {
            int index = (int) s[i] - 'a';
            if (!temp->list[index]) {
                temp->list[index] = newNode(s[i]);
            }
            temp = temp->list[index];
        }
        temp->cnt = 1;
    }
    void display(node *root) {
        if (!root)
            return;
        node *temp = root;
        cout << temp->info << " " << temp->cnt << '\n';
        for (int i = 0; i < 26; i++)
            display(temp->list[i]);
    }
    int countChild(node *temp, int &index) {
        int cnt = 0;
        for (int i = 0; i < 26; i++) {
            if (temp->list[i] != nullptr) {
                cnt++;
                index = i;
            }
        }
        return cnt;
    }
    string longestCommonPrefix(vector<string> &strs) {
        node *head = new node();
        for (int i = 0; i < strs.size(); i++) {
            insert(head, strs[i]);
        }
        node *curr = head;
        string lcp;
        int index;
        while (!curr->cnt && countChild(curr, index) == 1) {
            curr = curr->list[index];
            lcp += 'a' + index;
        }
        return lcp;
    }
};
